### DISCUSSION TITLE
Setup delay between disabling live preview and image capture?

### DISCUSSION THREAD

**Discussion by DominicWindisch** [photobooth-app/photobooth-app#161] [Q&A]
I'm using the gphoto2 backend for preview and image capture. Before capturing an image, preview has to be disabled. It appears that the Photobooth-App does not wait for the preview to stop before trying to capture an image resulting in "device busy" errors.

Is there an option to include a delay between disabling the preview and capturing an image?

**Comment by mgineer85** [comment]
Hi, 

currently there is no such exact option.
Which camera you use?
I have Canon 1100 and for this I need to enable the option "Gphoto2 Disable Viewfinder Before Capture" in the backend config. That would disable the liveview before capture. Maybe it helps to play around with that option and the option right below that one in the config.

What is the result with the current situation?
The image fails completely?
Can you post logs?

If you want to tinker around with delays or different config you could change the source to test. 
Find the installation folder using the logs. Look for a line like:
`2024-01-24 17:45:26,420 [    INFO] photobooth directory: D:\dev\photobooth\photobooth-app\photobooth (__main__.py:43)`
This is where the application is installed. Navigate there and you find the file gphoto2.py:
https://github.com/photobooth-app/photobooth-app/blob/main/photobooth/services/backends/gphoto2.py#L241

In the line linked above the capture command is sent.

**Reply by DominicWindisch** [reply]
I'm currently using an Canon EOS M3 for testing.

I have enabled that switch.

When I have live preview via gphoto2 enabled, no image is taken at all. The interface stays white for a couple of seconds and returns to the start page. During that period, the camera is released from gphoto2 (Indicated by its display turning on). After a few seconds the live preview resumes.

[latest.txt](https://github.com/photobooth-app/photobooth-app/files/14041674/latest.txt)

I'll have a look at the source later :)

**Comment by DominicWindisch** [comment]
It seems like complete overkill, but calling 
```
self._camera.exit()
self._camera.init()
```
instead of setting `viewfinder(0)` seems to work fine. I've tried all kinds of delays but my camera just wouldn't release resources. I'll likely investigate a bit deeper as to which parameter might be sufficient instead of this exit - init - monstrosity. Even though it works reasonably fast.

**Reply by mgineer85** [reply]
Oh Jeah that seems a bit overkill 🤣

maybe you can find something helpful here to trigger the camera different.

https://github.com/jim-easterbrook/python-gphoto2/tree/main/examples

quick google of your camera and gphoto2 gives mixed feelings about its support especially in combination with liveview.

**Reply by mgineer85** [reply]
Sometimes the camera mode can make a difference also

**Reply by DominicWindisch** [reply]
Yeah, thanks for the hints. I'll check what I can get to work. :)

I'll more than likely use a different camera for our event anyways. Got a few months left to tinker around 😁

**Reply by mgineer85** [reply]
You could also just add a usb webcam and use that for livestream to let the camera just take the high quality pics…

**Reply by DominicWindisch** [reply]
We're gonna use a 600D which should support viewfinder. Fingers crossed, I'll test in the next weeks.

I'm hesitant to use a second camera (but it would of course be any easy fallback option).

**Comment by DominicWindisch** [comment]
Using a Canon EOS 600D works tremendously 🤩

It might be useful to have a switch on whether to reconnect to the camera before the capture which appears to be the only way to use a camera that doesn't support viewfinder. I couldn't find a more graceful way yet to use the Canon EOS M3. But to be fair, it appears to be quite janky with gphoto2.

**Reply by mgineer85** [reply]
You mean a kind of reset before capture to handle m3 cam better? 
I think I can improve situation also by disable the live preview request at all if not well supported. That opens the way to use a second cam actually.

**Reply by DominicWindisch** [reply]
Yes, have the option to stop the preview by whatever means necessary. I'll tinker with that and try to find something more graceful than a hard reset.

The option to use a second backend for preview is there either way, why block the user from using the DSLR preview if he/she wishes to?

**Reply by mgineer85** [reply]
Block it because this would allow at least to capture hq still if preview fails and doesn’t mess with the cam that cannot deliver preview. … or gphoto fails to do so.

**Reply by DominicWindisch** [reply]
At least in case of my M3, the effect is essentially the same. If I disable preview, the camera does nothing until it gets the capture command. In the scenario with a reset, I have a fully functional preview in idle/countdown mode. The only downside is a 2 second blank period before the capture. 

The failure to capture only arises when skipping the reset. But I'd imagine people typically try the setup before and see if this reset needs to be enabled?

**Reply by mgineer85** [reply]
There is no difference because internally the preview ist still requested from the camera. This is not an issue for any of the non dslr backends but can be an issue for gphoto and digicamcontrol.
Will improve it and add more options.

**Reply by DominicWindisch** [reply]
BTW, from my tests as of now, you can add `Raspberry PI 4 / Raspberry PI OS / Bookworm / Canon EOS 600D` to the reference list. I've reverted my changes and this setup works perfectly fine. 

If anything, the delay for the preview seems a bit high (~2s), but I can't tell if that's a limitation of the camera itself. The preview from the M3 seems to have a way lower delay (~0.5s).

**Reply by mgineer85** [reply]
docs updated, once github actions pipeline finished it gets updated in some minutes.

**Comment by mgineer85** [comment]
You mean a lag in the stream?
For my canon 1100d it’s also more in the 0.5sec range.
The are no buffer involved anywhere, all data is sent out asap. Found some issues in gphoto2 repo about this also but leading to nothing unfortunately.

**Reply by DominicWindisch** [reply]
Same for me. The issue (I guess we found the same one) states that "real time" was possible at one point. I'll get back to you if I figure out anything useful...

**Reply by DominicWindisch** [reply]
When switching the 600D from Av mode into Tv mode, lag drastically reduces. It's now also in the range of 0.5s (which is totally fine). Still somewhat odd because the on-device monitor doesn't have any lag in either mode... :shrug: and the M3 also was in Av mode.
